warp10/src/main/java/io/warp10/script/binary/SUB.java

```diff
@@ -171,19 +164,31 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
       if (!(gts.getType() == TYPE.LONG || gts.getType() == TYPE.DOUBLE)) {
         throw new WarpScriptException(typeCheckErrorMsg);
       }
-      
-      double op = op1gts ? ((Number) op2).doubleValue() : ((Number) op1).doubleValue();
-      
-      for (int i = 0; i < n; i++) {
-       double value;
-       if (op1gts) {
-         value = ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue() - op;
-       } else {
-         value = op - ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue();
-       }
-       GTSHelper.setValue(result, GTSHelper.tickAtIndex(gts, i), GTSHelper.locationAtIndex(gts, i), GTSHelper.elevationAtIndex(gts, i), value, false);
-      }      
 
+
+      Number op = op1gts ? (Number) op2 : (Number) op1;
+
+      if (op instanceof Double || gts.getType() == TYPE.DOUBLE) {
+        for (int i = 0; i < n; i++) {
+          double value;
+          if (op1gts) {
+            value = ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue() - op.doubleValue();
+          } else {
+            value = op.doubleValue() - ((Number) GTSHelper.valueAtIndex(gts, i)).doubleValue();
```

Comment:
avoiding the function call op.doubleValue() would probably be more efficient, at least before inlining/JIT kicks in.

Reply:
Call to `op.doubleValue()` is mandatory because all we know is that it is a Number in the case `gts.getType() == TYPE.DOUBLE`.

Reply:
yes but you don't have to do it in the loop!